### ISSUE TITLE
Can we fetch patch embeddings for SAM encoder ? If yes, then how ?

### ISSUE THREAD

**Issue by yogendra-yatnalkar** [facebookresearch/segment-anything#588]
As SAM encoder is ViT based model, can I get patch embeddings out of it ?

---

**Context:** 
SAM encoder uses image input of **1024x1024** and patch size of **16x16**. The encoder also consists of a neck module which has few Conv2D layers. The output shape of the encoder post the neck module is: **64x64x256.** 

Lets say, I want patch embeddings of the top-left 4 patches, its easily possible in models like **ViT and DinoV2 (again, ViT based)**, but I am not able to understand on how to do the same over here. 

---

Can some please guide me on this ? Thanks a lot in advance.

**Comment by heyoeyo** [comment]
Yes it's possible to directly access the output of the image encoder. The embeddings are available from the `SamPredictor` object, under the `.features` attribute. The simplest way I can think of would be something like:

```
import cv2
from segment_anything import SamPredictor, sam_model_registry

# Load the image that you want to process
example_image = cv2.imread("path/to/image.jpg")

# This comes from the 'Getting Started' sample code provided on the repo page
sam = sam_model_registry["vit_b"](checkpoint="path/to/sam_model_vit-b.pth")
predictor = SamPredictor(sam)
predictor.set_image(example_image)

# *** This is how you can get the patch embeddings ***
patch_embeddings = predictor.features
# patch_embeddings.shape gives: torch.Size([1, 256, 64, 64])
```

If you wanted the top-left 4 patches (assuming 2x2 layout), you could do something like:
`top_4_patches = patch_embeddings[:,:,:2,:2]`

Alternatively, you could also just instantiate a copy of the [image encoder class](https://github.com/facebookresearch/segment-anything/blob/main/segment_anything/modeling/image_encoder.py) itself and run it's `.forward(...)` method to directly generate the patch embeddings. However, this would actually require a bit of messing about with how the model weights are loaded, in addition to handling the image preprocessing manually, so probably not as easy!

**Comment by yogendra-yatnalkar** [comment]
Hi @heyoeyo thanks for your response. For the second part, I have already figured out how to run the forward method of SAM encoder which takes input of: **Batch_sizex3x1024x1024** image dimensions and output embedding of dim: **batch_sizex256x64x64**

As far as the embedding goes, it comes out of a neck module which has 2 conv2d layers, hence I got confused in understanding if it can be used as patch_embeddings or not. 

But thanks to you, I rechecked the encoder code once and I believe it should work. The conv-neck module is only to reduce the dimensions of the embedding shape but the patches are still remaining consistent. Thanks a lot for your kind reply. 

---
Just putting down my new understanding, feel free to correct me if required: 

If patch size is **16x16** and image size is **1024x1024**, the number of patches are actually **64*64** only and even our output shape is: **256x64x64** so this makes sense now.

**Comment by heyoeyo** [comment]
That sounds right. It's a bit confusing looking at the code, because there is a [patch_embed](https://github.com/facebookresearch/segment-anything/blob/6fdee8f2727f4506cfbbe553e23b895e27956588/segment_anything/modeling/image_encoder.py#L107) component which performs the first processing step that converts the input image into the 64x64 patches. However, that step gives different dimensions depending on the model used: 768x64x64 (vit-b), 1024x64x64 (vit-l) or 1280x64x64 (vit-h), and is only a small part of all the computation the model was trained to do.

Like you mentioned, there's a neck component that reduces the 768/1024/1280 dimension down to 256. There might be some benefit to using the tensor from just before the neck processing, depending on what you're doing with it. As far as I understand, the smallest size dimension (768) was chosen because it matches the 'dimensionality' of the underlying data, that is a 16x16 group of RGB values (16x16x3 = 768), whereas a size of 256 has to be 'throwing away' information, in some sense.

**Comment by ByungKwanLee** [comment]
https://github.com/ByungKwanLee/Full-Segment-Anything addresses the ciritical issues of SAM, which supports batch-input on the full-grid prompt (automatic mask generation) **with post-processing**: removing duplicated or small regions and holes, under **flexible input image size**

**Comment by yogendra-yatnalkar** [comment]
Hi @ByungKwanLee, thanks a lot for sharing your work. For accommodating various different image sizes, you have added 1 parameter in the  **build_sam** code which is: **custom_img_size**. I do understand on high-level what would be happening  ahead but wanted to check with you few questions I had in my mind: 

- If we reduce the image size, will it drastically degrade the performance ?
- If yes, can you please share your experiments around this ?

PS: your work looks amazing, thanks again for sharing and thank you in advance.

**Comment by ByungKwanLee** [comment]
Thanks for your interest in full-segment-anything. Actually, I only modified the specific part of the original SAM code in order to use custom image size, for your purpose, rather than fixed 1024 resolution. In other words, I did not re-train SAM architecture to cover dynamic resolution size, but conduct slight modifications in the code-level. Therefore, it is unfortunately impossible to provide quantitative results, under various resolution, compred by those of SAM. However, it is definitely sure that image-encoder is highly important part to understand the image before decoding and its performance is very dependent with the image resolution size, hence performance degradation is naturally happening once we down-size the image.
